### PR TITLE
[MODULES] Switch to Go Modules

### LINKED ISSUE #84: [PROPOSAL] Switch to Go Modules
As part of the preparation for Terraform v0.12, we would like to migrate all providers to use [Go Modules](https://github.com/golang/go/wiki/Modules). We plan to continue checking dependencies into vendor/ to remain compatible with existing tooling/CI for a period of time, however go modules will be used for management. Go Modules is the official solution for the go programming language, we understand some providers might not want this change yet, however we encourage providers to begin looking towards the switch as this is how we will be managing all Go projects in the future. Would maintainers please react with :+1: for support, or :-1: if you wish to have this provider omitted from the first wave of pull requests. If your provider is in support, we would ask that you avoid merging any pull requests that mutate the dependencies while the Go Modules PR is open (in fact a total codefreeze would be even more helpful), otherwise we will need to close that PR and re-run `go mod init`. Once merged, dependencies can be added or updated as follows:

```
$ GO111MODULE=on go get github.com/some/module@master
$ GO111MODULE=on go mod tidy
$ GO111MODULE=on go mod vendor
```

GO111MODULE=on might be unnecessary depending on your environment, this example will fetch a module @ master and record it in your project's go.mod and go.sum files. It's a good idea to tidy up afterward and then copy the dependencies into vendor/. To remove dependencies from your project, simply remove all usage from your codebase and run:

```
$ GO111MODULE=on go mody tidy
$ GO111MODULE=on go mod vendor
```

Thank you sincerely for all your time, contributions, and cooperation!

### PR DESCRIPTION
**Pull request by appilon**
Closes #84

### CODE REVIEW

**Comment by appilon** [comment]
I pinned the huawei cloud sdk back to a commit before a breaking change was introduced.

File: go.sum [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+github.com/mitchellh/panicwrap v0.0.0-20161208170302-ba9e1a65e0f7 h1:+PBI9A4rLQJlch3eQI/RkTY2HzX+bl2lPUf3goenBxs=
+github.com/mitchellh/panicwrap v0.0.0-20161208170302-ba9e1a65e0f7/go.mod h1:QuAqW7/z+iv6aWFJdrA8kCbsF0OOJVKCICqTcYBexuY=
+github.com/mitchellh/prefixedio v0.0.0-20151214002211-6e6954073784 h1:+DAetXqxv/mSyCkE9KBIYOZs9b68y7SUaDCxQMRjA68=
+github.com/mitchellh/prefixedio v0.0.0-20151214002211-6e6954073784/go.mod h1:kB1naBgV9ORnkiTVeyJOI1DavaJkG4oNIq0Af6ZVKUo=
+github.com/mitchellh/reflectwalk v0.0.0-20170726202117-63d60e9d0dbc h1:gqYjvctjtX4GHzgfutJxZpvZ7XhGwQLGR5BASwhpO2o=
+github.com/mitchellh/reflectwalk v0.0.0-20170726202117-63d60e9d0dbc/go.mod h1:mSTlrgnPZtwu0c4WaC2kGObEpuNDbx0jmZXqmk4esnw=
+github.com/nu7hatch/gouuid v0.0.0-20131221200532-179d4d0c4d8d h1:VhgPp6v9qf9Agr/56bj7Y/xa04UccTW04VP0Qed4vnQ=
+github.com/nu7hatch/gouuid v0.0.0-20131221200532-179d4d0c4d8d/go.mod h1:YUTz3bUH2ZwIWBy3CJBeOBEugqcmXREj14T+iG/4k4U=
+github.com/oklog/run v1.0.0 h1:Ru7dDtJNOyC66gQ5dQmaCa0qIsAUFY3sFpK1Xk8igrw=
+github.com/oklog/run v1.0.0/go.mod h1:dlhp/R75TPv97u0XWUtDeV/lRKWPKSdTuV0TZvrmrQA=
```
Comment (radeksimko) [inline go.sum:164]: It looks like this dependency is causing some issues for `go mod tidy`:

```
go: finding github.com/dnaeon/go-vcr v0.0.0-20170218072653-87d4990451a8
go: github.com/oklog/run@v1.0.0: unknown revision v1.0.0
go: finding github.com/Azure/azure-sdk-for-go v10.3.0-beta+incompatible
```

Do you mind `rm go.sum` and re-running `go mod tidy`?

Comment (appilon) [reply go.sum:164]: How are you able to get that? Hmm makes me think modules still has some inconsistent behavior

Comment (appilon) [reply go.sum:164]:
```
~/git/go/src/github.com/terraform-providers/terraform-provider-huaweicloud
❯ GO111MODULE=on go mod tidy
go: finding github.com/smartystreets/goconvey v0.0.0-20190222223459-a17d461953aa
go: finding github.com/huaweicloud/golangsdk v0.0.0-20190218081629-94fd0333a106
go: finding github.com/hashicorp/terraform v0.11.12-beta1.0.20190227065421-fc531f54a878
go: finding github.com/hashicorp/go-tfe v0.3.10
go: finding golang.org/x/sys v0.0.0-20190228124157-a34e9553db1e
go: finding gopkg.in/yaml.v2 v2.0.0-20170812160011-eb3733d160e7
go: downloading github.com/hashicorp/terraform v0.11.12-beta1.0.20190227065421-fc531f54a878
go: downloading github.com/huaweicloud/golangsdk v0.0.0-20190218081629-94fd0333a106
go: downloading gopkg.in/yaml.v2 v2.0.0-20170812160011-eb3733d160e7
go: downloading github.com/smartystreets/goconvey v0.0.0-20190222223459-a17d461953aa
go: downloading golang.org/x/sys v0.0.0-20190228124157-a34e9553db1e

~/git/go/src/github.com/terraform-providers/terraform-provider-huaweicloud 8s
❯ git status
On branch go-modules-2019-03-01
nothing to commit, working tree clean
```

Comment (appilon) [reply go.sum:164]: I really don't like that when I run `go mod tidy` with the go.sum existing, it is unchanged, but delete it and re-run it changes...

Comment (radeksimko) [reply go.sum:164]: > I really don't like that when I run go mod tidy with the go.sum existing, it is unchanged, but delete it and re-run it changes...

That is odd indeed. I assume there are still some bugs in go modules which get triggered by some edge-case-y, hard-to-reproduce situations when something happens in the cache. 😞 

Either way I always make sure that I test in a clean environment without any cache etc. when I test these PRs, specifically in `golang:1.11.5` in Docker. The only downside of this approach is just long wait until all dependencies get downloaded from scratch inside the container.

**Review by radeksimko — CHANGES_REQUESTED** [review]
The code looks good and I can confirm that `go mod tidy` & `go mod vendor` are both no-op with latest changes, but I'll have to put this on hold until we have valid credentials to test this with.
